kani-driver/src/concrete_playback.rs

```diff
@@ -256,156 +281,127 @@ struct UnitTest {
+                    "WARNING: Unable to extract concrete values from multiple failing assertions. Skipping property `{}` with description `{}`.",
+                    property.property, property.description,
                 );
             } else {
-                *have_parsed_assert_fail = true;
+                *extracted_assert_fail = true;
                 println!(
-                    "INFO: Parsing concrete values from property `{prop}` with description `{desc}`."
+                    "INFO: Parsing concrete values from property `{}` with description `{}`.",
+                    property.property, property.description,
                 );
-                let trace_arr = result_val["trace"].as_array().with_context(|| {
-                    format!(
-                        "Expected this CBMC result trace to be an array: {}",
-                        result_val["trace"]
-                    )
-                })?;
-                for trace_entry in trace_arr {
-                    parse_trace_entry(trace_entry, concrete_vals)
-                        .context("Failure in trace assignment expression:")?;
+                if let Some(trace) = &property.trace {
+                    for trace_item in trace {
+                        let concrete_val_opt = extract_from_trace_item(trace_item)
+                            .context("Failure in trace assignment expression:")?;
+                        if let Some(concrete_val) = concrete_val_opt {
+                            concrete_vals.push(concrete_val);
+                        }
+                    }
                 }
             }
-        } else if !prop_is_assert && status_is_failure {
+        } else if !property_is_assert && status_is_failure {
             println!(
-                "WARNING: Unable to parse concrete values from failing non-assertion checks. Skipping property `{prop}` with description `{desc}`."
+                "WARNING: Unable to extract concrete values from failing non-assertion checks. Skipping property `{}` with description `{}`.",
+                property.property, property.description,
             );
         }
-        Ok(())
+        Ok(concrete_vals)
     }
 
-    /// The third-level CBMC output parser. This extracts individual bytes from kani::any_raw calls.
-    fn parse_trace_entry(trace_entry: &Value, concrete_vals: &mut Vec<ConcreteVal>) -> Result<()> {
-        if let (
-            Some(step_type),
-            Some(lhs),
-            Some(func),
-            Some(bit_concrete_val),
-            Some(interp_concrete_val),
-            Some(width_u64),
-        ) = (
-            trace_entry["stepType"].as_str(),
-            trace_entry["lhs"].as_str(),
-            trace_entry["sourceLocation"]["function"].as_str(),
-            trace_entry["value"]["binary"].as_str(),
-            trace_entry["value"]["data"].as_str(),
-            trace_entry["value"]["width"].as_u64(),
-        ) {
```

Comment:
Nice! You don't need `step_type` anymore?

Reply:
Nope, we're still using it later in that function. However, because `step_type` is not an option, we don't need to extract it in the `if let`.